source/docs/advanced/features/external-logins.md

```diff
@@ -0,0 +1,314 @@
+
+To create  a provider  you need to register a key and a function with the `auth-external-login` module.
+
+The key will be used in the redirect url (like "google" in `/external-login/google` and `/external-login/google/callback`) and the function will be called to create the provider.
+
+The function will be called with 3 arguments:
+* The current request
+* The externalLoginHandler
+* The redirectUrl
+
+The provider returned by the registered function should have 2 middlewares `requestLoginMiddleware` and `loginCallbackMiddleware`.
+
+The `requestLoginMiddleware` will be called both when the user clicks on the "Login with Provider" button and when the provider redirects back to the site. The role of `requestLoginMiddleware` is to first redirect the user to the external site to authenticate and later when the user is redirected back to the site (from the external system) to extract the user information.
+
+The `loginCallbackMiddleware` is called after the user is redirected back to the site (from the external system) and also after `requestLoginMiddleware`'s second call by this time `requestLoginMiddleware` would have hopefully extracted all needed user fields and made them available for `loginCallbackMiddleware`. (We/passort does it by exposing a field called `user` on the request object itself). Here is where the external login handler is used to log the the user in.
+
+```js
+import configService from "server/core/config/configService";
+import { getCurrentShopConfig } from "server/core/config/getShopConfig";
+import { getCurrentShopConfig } from "server/core/config/getShopConfig";
+import customConfigProvider from "path_to_your_customConfigProvider";
+import { registerExternalLoginProvider } from "auth-external-login/server/providers";
+
+configService.append(customConfigProvider);
+
+const makeCustomProvider = (request, externalLoginHandler, callbackUrl) => {
+  const providerConfig = request.config.customProviderConfig;
+  if(!providerConfig.clientId || !providerConfig.clientSecret) {
+    throw new Error(`Custom provider is not properly configured. clientId and clientSecret are required!`);
+  }
+  const shopUrl = getCurrentShopConfig(req.config).url; // frequently needed to keep current store selected upon redirection
+  return new CustomProvider(externalLoginHandler, providerConfig, {
+    shopUrl,
+    callbackUrl,
+  });
+};
+
+registerExternalLoginProvider("custom", makeCustomProvider);
+```
+
+#### The LoginProviderInerface
+
+To make creating a provider easier, `LoginProviderInerface` can be implemented. By implementing `LoginProviderInerface` you have to implement the 2 aforementioned middleware and you can pick what inputs your constructor takes. You will then register the maker function mentioned above to extract the data needed by your provider and return an instance of it.
```

Comment:
```suggestion
To make creating a provider easier, `LoginProviderInterface` can be implemented. By implementing `LoginProviderInterface` you have to implement the 2 aforementioned middleware and you can pick what inputs your constructor takes. You will then register the maker function mentioned above to extract the data needed by your provider and return an instance of it.
```

Reply:
The role of this chapter is unclear to me. Detailling here and after the full detailed reasoning adds too much complexity in my opinion. 

Why not just telling an integrator what to add and its simple role ?
- Handling a new provider and registering it
- Handling a backend that is not implemented and registering it

Reply:
its our job is to help them and we have these classes/interfaces that could help. Thats why I mention them. Otherwise they can just return an object with the middlewares implemented no problem and all will work fine

Reply:
I think it is better to just tell them to implement the interface then, as it will both enforce the implementation AND send them direct warning while upgrading FC